packages/core/src/components/button/abstractButton.tsx

```diff
@@ -40,31 +50,65 @@ export abstract class AbstractButton<T> extends React.Component<React.HTMLProps<
 
     public abstract render(): JSX.Element;
 
-    protected onKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
-        switch (e.which) {
-            case Keys.SPACE:
-            case Keys.ENTER:
-                e.preventDefault();
-                if (e.which !== this.currentKeyDown) {
-                    this.setState({ isActive: true });
-                }
-                break;
-            default:
-                break;
+    protected getCommonButtonProps() {
+        const disabled = this.props.disabled || this.props.loading;
+
+        const className = classNames(
+            Classes.BUTTON,
+            {
+                [Classes.ACTIVE]: this.state.isActive,
+                [Classes.DISABLED]: disabled,
+                [Classes.LOADING]: this.props.loading,
+            },
+            Classes.iconClass(this.props.iconName),
+            Classes.intentClass(this.props.intent),
+            this.props.className,
+        );
+
+        return {
+            className,
+            disabled,
+            onClick: disabled ? undefined : this.props.onClick,
+            onKeyDown: this.handleKeyDown,
+            onKeyUp: this.handleKeyUp,
+            ref: this.refHandlers.button,
+        };
+    }
+
+    protected handleKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
+        if (isKeyboardClick(e.which)) {
+            e.preventDefault();
+            if (e.which !== this.currentKeyDown) {
+                this.setState({ isActive: true });
+            }
         }
         this.currentKeyDown = e.which;
     }
 
-    protected onKeyUp = (e: React.KeyboardEvent<HTMLElement>) => {
-        switch (e.which) {
-            case Keys.SPACE:
-            case Keys.ENTER:
-                this.setState({ isActive: false });
-                this.buttonRef.click();
-                break;
-            default:
-                break;
+    protected handleKeyUp = (e: React.KeyboardEvent<HTMLElement>) => {
+        if (isKeyboardClick(e.which)) {
+            this.setState({ isActive: false });
+            this.buttonRef.click();
         }
         this.currentKeyDown = null;
     }
+
+    protected renderChildren(): React.ReactNode {
+        const { children, loading, rightIconName, text } = this.props;
+        const iconClasses = classNames(Classes.ICON_STANDARD, Classes.iconClass(rightIconName), Classes.ALIGN_RIGHT);
+        return [
+            maybeRender(loading, <Spinner className="pt-small pt-button-spinner" key="spinner" />),
+            maybeRender(text != null, <span key="text">{text}</span>),
+            children,
+            maybeRender(rightIconName != null, <span className={iconClasses} key="icon" />),
+        ];
+    }
+}
+
+function maybeRender(condition: boolean, element: JSX.Element, defaultValue: JSX.Element = undefined) {
+    return condition ? element : defaultValue;
```

Comment:
`maybeRender` seems the correct name to me. if anything it should live in utils, next to `safeInvoke`. no need to add this static function to AbstractComponent.

but i'm still not convinced that this is strictly better than the inline ternary ifs i had before. more complex / multiline expressions are best implemented as one-off functions (which we already do on a per-component basis).